compiler/resolution/preFold.cpp

```diff
@@ -233,42 +239,73 @@ static Expr* preFoldPrimOp(CallExpr* call) {
 
   case PRIM_GATHER_TESTS: {
     int  totalTest = 0;
+    Type* testType = call->get(1)->getValType();
     forv_Vec(FnSymbol, fn, gFnSymbols) {
       if (fn->throwsError()) {
         ModuleSymbol *mod = fn->getModule();
         if (mod->modTag == MOD_USER) {
           if (fn->numFormals() == 1) {
             if (!fn->hasFlag(FLAG_GENERIC) && fn->instantiatedFrom == NULL) {
               const char* name = astr(fn->name);
-
-              // temporarily add a call to try resolving.
-              CallExpr* tryCall = new CallExpr(name);
-              // Add our new call to the AST temporarily.
-              call->getStmtExpr()->insertAfter(tryCall);
-
-              // copy actual args into tryCall.
-              for_actuals(actual, call) {
-                tryCall->insertAtTail(actual->copy());
-              }
-
-              // Try to resolve it.
-              if (tryResolveCall(tryCall)) {
+              resolveSignature(fn);
+              if(isSubTypeOrInstantiation(fn->getFormal(1)->type, testType,call)) {
                 totalTest++;
+                CallExpr* newCall = new CallExpr(PRIM_CAPTURE_FN_FOR_CHPL, new UnresolvedSymExpr(name));
+                fn->defPoint->getStmtExpr()->insertAfter(newCall);
+                testCaptureVector.push_back(createFunctionAsValue(newCall));
+                newCall->remove();
+                testNameIndex[name] = totalTest;
               }
-
-              // remove the call from the AST
-              tryCall->remove();
             }
           }
         }
-        
       }
     }
     retval=new SymExpr(new_IntSymbol(totalTest));
     call->replace(retval);
     break;
   }
 
+  case PRIM_GET_TEST_BY_INDEX: {
+    int64_t index    =        0;
+    
+    if (call->numActuals() == 0) {
+      USR_FATAL(call, "illegal call of type");
+    }
+
+    if (call->numActuals() > 1) {
+      USR_FATAL(call, "too many arguments to type index expression");
+    }
+
+    if (!get_int(call->get(1), &index)) {
+      USR_FATAL(call, "illegal type index expression");
+    }
+
+    if (index <= 0 || index > (int64_t)testCaptureVector.size()) {
+      USR_FATAL(call, "type index expression '%i' out of bounds", index);
```

Comment:
I would change this to `index '%i' out of bounds, expected to be between '%i' and '%i'`

Reply:
I have made the changes